app/services/form.rb

```diff
@@ -0,0 +1,38 @@
+class Form
+  def initialize(json:, schema:, struct:)
+    @json = json || "{}"
+    @schema = schema
+    @struct = struct
+    @errors = []
+  end
+
+  def invalid?
+    @schema.call(params).failure?
+  end
+
+  def errors
+    @schema.call(params).messages(full: true)
+  end
+
+  def to_h
+    @struct.new(
+      @schema.call(params).to_h
+    )
+  end
+
+  def add_error(error)
```

Comment:
Not used for now but it's useful in case we want from the service to return added errors